flat_mutation_reader.cc

```diff
@@ -912,3 +915,48 @@ class generating_reader final : public flat_mutation_reader::impl {
 flat_mutation_reader make_generating_reader(schema_ptr s, std::function<future<mutation_fragment_opt> ()> get_next_fragment) {
     return make_flat_mutation_reader<generating_reader>(std::move(s), std::move(get_next_fragment));
 }
+
+bool mutation_fragment_stream_validator::operator()(const dht::decorated_key& dk) {
+    // FIXME: Validate fragment position monotonicity
+    //        https://github.com/scylladb/scylla/issues/4804
+    return true;
+}
+
+mutation_fragment_stream_validator::mutation_fragment_stream_validator(const schema& s)
+    : _schema(s)
+    , _prev_kind(mutation_fragment::kind::partition_end)
+    , _prev_pos(position_in_partition::end_of_partition_tag_t{})
+{ }
+
+mutation_fragment_stream_validator::~mutation_fragment_stream_validator() {
+    if (_prev_kind != mutation_fragment::kind::partition_end) {
+        on_internal_error(fmr_logger, format("Stream ended with unclosed partition: {}", _prev_kind));
+    }
+}
+
+bool mutation_fragment_stream_validator::operator()(const mutation_fragment& mv) {
+    auto kind = mv.mutation_fragment_kind();
+    auto pos = mv.position();
+    bool valid;
```

Comment:
I think it would be best to initialize this, to avoid UB due to mistakes.

Reply:
I beg to differ. We actually want the compiler to warn us if there's a path that doesn't set it.  Initializing it will prevent that. Ideally it could be initialized to `false` only for non-debug builds.

Reply:
Does it warn for that? I though these bugs can only be caught at runtime.
If there is a warning for that then sure, no need to initialize it defensively.

Reply:
Well, `-Wmaybe-uninitialized` does that but we disable it.  IIRC it produces too many false warnings.  I'll add the initialization then.

Reply:
I think it was disabled due to bad interactions with `std::optional`. I'm not sure anymore. I think, from time-to-time we should go through our disabled warning list and see if they can be re-enabled.